File: src/medium/t5/solve.go
package main

import (
	"sort"
)

func threeSum(nums []int) [][]int {
	// 为了保证不重复，先从小到大排序
	sort.Ints(nums)
	// [-4,-1,-1,0,1,2]
	var results = [][]int{}
	n := len(nums)
	// 固定 最小的数， 后两个数指针游动
	for i := 0; i < n-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			// 保证不重复
			continue
		}
		// 三数相加等于0，如果最小的三个数就大于0,没有更小的数了
		if nums[i]+nums[i+1]+nums[i+2] > 0 {
			return results
		}
		// 如果最小的一个值加最大的两个值都小于0，终止循环
		if nums[i]+nums[n-2]+nums[n-1] < 0 {
			continue
		}
		//  第三个数的游标指针
		var k = n - 1
		// 对第二个数的指针进行滑动，每次前进一格
		var j = i + 1
		for j < k {
			var sum = nums[j] + nums[k] + nums[i]
			// 如果三个数相加大于0，将最大的数往左移动一位
			if sum > 0 {
				k--

			} else if sum < 0 {
				j++
			} else {
				// 否则就等于0了
				results = append(results, []int{nums[i], nums[j], nums[k]})
				// 继续下一个指针数
				for j < k && nums[j+1] == nums[j] {
					j++
				}
				for j < k && nums[k-1] == nums[k] {
					k--
				}
				j++
				k--
			}
		}
	}
	return results
}

File: src/medium/t5/solve.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        result = []
        nums.sort()
        for i in range(0, n-2):
            first = nums[i]
            if i != 0 and nums[i] == nums[i-1]:
                continue
            if first + nums[i+1]+nums[i+2] > 0:
                return result
            if first + nums[n-1] + nums[n-2] < 0:
                continue
            j, k = i+1, n - 1
            while j < k:
                sum = first+nums[j]+nums[k]
                if sum > 0:
                    k -= 1
                elif sum < 0:
                    j += 1
                else:
                    result.append([first, nums[j], nums[k]])
                    while j < k and nums[j+1] == nums[j]:
                        j += 1
                    while j < k and nums[k-1] == nums[k]:
                        k -= 1
                    j += 1
                    k -= 1
        return result


nums = [-1, 0, 1, 2, -1, -4]
s = Solution()
s.threeSum(nums)
